### PR TITLE
Release - 3.0.0

### PR DESCRIPTION
**Pull request by supercid**
## Description
* Update marketing permission to Nosto in real-time when newsletter subscription is changed
* Support adding multiple products to cart from Nosto recommendations
* Introduce a cli command for connecting Nosto account via command line
* Improve performance for generating tagging (@hostep)
* Fix the issue with product building when no custom fields are found (@hostep)
* Improve error handling for Nosto dashboard in store admin area
* Code style fixes & refactoring
* Set product into Magento registry before fetching Yotpo reviews
* Add Yotpo reviews feature
* Add customer group pricing variation support
* Support using same nosto email widget snippet for multiple nosto accounts

## How Has This Been Tested?
- [x] Deployed to Plugin test instance
- [x] Robot tests have passed
## Checklist:

- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.
- [x] I have assigned the correct milestone or created one if non existent.
- [x] I have correctly labeled this pull request.
- [x] I have requested a review from at least 2 reviewers

### CODE REVIEW

File: Block/Category.php [hunk 1]
```diff
@@ -101,13 +105,12 @@ private function getNostoCategory()
     /**
      * Returns the HTML to render categories
      *
-     * @return MarkupableString
+     * @return string
      */
-    public function getAbstractObject()
+    public function toHtml()
```
Comment (hannunosto) [inline Block/Category.php]: No need to override the `toHtml`. Keep the `getAbstractObject` method and return `$category`

File: Block/PageType.php [hunk 2]
```diff
@@ -83,14 +57,19 @@ public function __construct(
...
     }
 
-    public function getAbstractObject()
+    /**
+     * Returns the HTML to render PageTypes
+     *
+     * @return string
+     * @suppress PhanUndeclaredClassMethod
+     */
+    public function toHtml()
```
Comment (hannunosto) [inline Block/PageType.php]: Same thing than with the category. No need to override the `toHtml`. Keep the `getAbstractObject` method and return the object

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,18 @@
...
+* Support adding multiple products to cart from Nosto recommendations
+* Introduce a cli command for connecting Nosto account via command line
+* Improve performance for generating tagging (@hostep)
+* Fix the issue with product building when no custom fields are found (@hostep)
+* Improve error handling for Nosto dashboard in store admin area
+* Code style fixes & refactoring
+* Set product into Magento registry before fetching Yotpo reviews
+* Add Yotpo reviews feature
+* Add customer group pricing variation support
+* Support using same nosto email widget snippet for multiple nosto accounts
```
Comment (hannunosto) [inline CHANGELOG.md]: nosto -> Nosto

File: Controller/Frontend/Cart.php [hunk 4]
```diff
@@ -111,38 +110,41 @@ public function execute()
...
-
         return $this->_redirect($redirectUrl);
     }
 
     /**
      * Resolves the cart (quote) by the given hash
      *
      * @param $restoreCartHash
-     * @return Quote|null
+     * @return \Magento\Quote\Api\Data\CartInterface|null
```
Comment (hannunosto) [inline Controller/Frontend/Cart.php]: Import class

File: Controller/Frontend/Cart.php [hunk 5]
```diff
@@ -111,38 +110,41 @@ public function execute()
...
      * @throws NostoException
      * @throws \Magento\Framework\Exception\NoSuchEntityException
```
Comment (hannunosto) [inline Controller/Frontend/Cart.php]: Import class

File: Helper/Variation.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    public function __construct(GroupRepository $groupRepository)
+    {
+        $this->groupRepository = $groupRepository;
+    }
+
+    /**
+     * @return \Magento\Customer\Api\Data\GroupInterface|null
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
```
Comment (hannunosto) [inline Helper/Variation.php]: Import all classes. Same thingy below as well

File: Model/Cart/Item/Grouped.php [hunk 7]
```diff
@@ -37,33 +37,57 @@
...
+        $this->productRepository = $productRepository;
+    }
+
     /**
      * Returns the name of the product. Grouped products will have their parent's name prepended to
      * their name.
      *
      * @param Item $item the ordered item
      * @return string|null the name of the product
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
```
Comment (hannunosto) [inline Model/Cart/Item/Grouped.php]: Import class

File: Model/Cart/Item/Grouped.php [hunk 8]
```diff
@@ -37,33 +37,57 @@
...
-
         return $name;
     }
+
+    /**
+     * Query the product id and returns the Product Object
+     *
+     * @param $productId
+     * @return \Magento\Catalog\Api\Data\ProductInterface|mixed
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
```
Comment (hannunosto) [inline Model/Cart/Item/Grouped.php]: Import classes

File: Model/Order/Builder.php [hunk 9]
```diff
@@ -110,7 +104,11 @@ public function build(Order $order)
                     $nostoOrder->setCreatedAt($orderCreatedDate);
                 }
             }
-            $nostoOrder->setPaymentProvider($order->getPayment()->getMethod());
+            if ($order->getPayment()) {
```
Comment (hannunosto) [inline Model/Order/Builder.php]: Should we check the return type?

File: Model/Order/Item/Builder.php [hunk 10]
```diff
@@ -170,14 +182,12 @@ public function buildSkuId(Item $item)
             //An item with bundle product and group product may have more than 1 child.
             //But configurable product item should have max 1 child item.
             //Here we check the size of children, return only if the size is 1
-            if (count($children) == 1
-                && array_key_exists(0, $children)
+            /** @var Item[] $children */
+            if (array_key_exists(0, $children) && $children[0] instanceof Item
```
Comment (hannunosto) [inline Model/Order/Item/Builder.php]: The second condition should be on own line as well

File: Model/Product/QueueRepository.php [hunk 11]
```diff
@@ -212,7 +208,7 @@ public function getAll()
     /**
      * @param SearchCriteriaInterface $searchCriteria
      *
-     * @return QueueSearchResults
+     * @return \Magento\Framework\Api\Search\SearchResult
```
Comment (hannunosto) [inline Model/Product/QueueRepository.php]: Import class

File: Helper/Variation.php (hunk 6 above)
Comment (supercid) [reply Helper/Variation.php]: Imported, not shown due to the comment on other line

File: Controller/Frontend/Cart.php (hunk 5 above)
Comment (supercid) [reply Controller/Frontend/Cart.php]: Imported

File: Controller/Frontend/Cart.php (hunk 4 above)
Comment (supercid) [reply Controller/Frontend/Cart.php]: Imported

File: CHANGELOG.md (hunk 3 above)
Comment (supercid) [reply CHANGELOG.md]: Fixed

File: Block/PageType.php (hunk 2 above)
Comment (supercid) [reply Block/PageType.php]: Reverted

File: Block/Category.php (hunk 1 above)
Comment (supercid) [reply Block/Category.php]: Reverted

File: Model/Cart/Item/Grouped.php (hunk 7 above)
Comment (supercid) [reply Model/Cart/Item/Grouped.php]: Imported

File: Model/Cart/Item/Grouped.php (hunk 8 above)
Comment (supercid) [reply Model/Cart/Item/Grouped.php]: Imported

File: Model/Order/Builder.php (hunk 9 above)
Comment (supercid) [reply Model/Order/Builder.php]: Checked for `Magento\Sales\Model\Order\Payment`

File: Model/Order/Item/Builder.php (hunk 10 above)
Comment (supercid) [reply Model/Order/Item/Builder.php]: Fixed

File: Model/Product/QueueRepository.php (hunk 11 above)
Comment (supercid) [reply Model/Product/QueueRepository.php]: Imported

File: Console/Command/NostoAccountConnectCommand.php [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    /**
+     * Check if required arguments passed by command line are present,
+     * if not, will ask for the remaining parameters.
+     *
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return array of Token objects
+     * @throws \Nosto\NostoException
```
Comment (hannunosto) [inline Console/Command/NostoAccountConnectCommand.php]: Import class

File: Controller/Adminhtml/Account/Create.php [hunk 13]
```diff
@@ -114,8 +114,11 @@ public function __construct(
 
     /**
      * @return Json
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Comment (hannunosto) [inline Controller/Adminhtml/Account/Create.php]: Import class

File: Controller/Adminhtml/Account/Index.php [hunk 14]
```diff
@@ -67,6 +67,8 @@ public function __construct(
 
     /**
      * @return Page | Redirect
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NotFoundException
```
Comment (hannunosto) [inline Controller/Adminhtml/Account/Index.php]: Import classes

File: Controller/Export/Order.php [hunk 15]
```diff
@@ -79,6 +79,7 @@ public function __construct(
      * @param int $limit
      * @param int $offset
      * @return \Nosto\Object\AbstractCollection|\Nosto\Object\Order\OrderCollection
+     * @throws \Nosto\NostoException
```
Comment (hannunosto) [inline Controller/Export/Order.php]: Import classes

File: Controller/Export/Order.php [hunk 16]
```diff
@@ -90,6 +91,7 @@ public function buildExportCollection(Store $store, $limit = 100, $offset = 0)
      * @param Store $store
      * @param int $id
      * @return \Nosto\Object\AbstractCollection|\Nosto\Object\Order\OrderCollection
+     * @throws \Nosto\NostoException
```
Comment (hannunosto) [inline Controller/Export/Order.php]: Import classes

**Pull request merged**